tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/source_files/source_files_container_test.ts

```diff
@@ -0,0 +1,237 @@
+    });
+    store.overrideSelector(getFocusedSourceLineSpec, {
+      host_name: 'localhost',
+      file_path: '/home/user/main.py',
+      lineno: 3,
+    });
+    store.refreshState();
+    fixture.detectChanges();
+    await fixture.whenStable();
+
+    let fileLabelElement = fixture.debugElement.query(By.css('.file-label'));
+    expect(fileLabelElement.nativeElement.innerText).toBe('/home/user/main.py');
+    expect(loadMonacoSpy).toHaveBeenCalledTimes(1);
+    expect(editorSpy.setValue).not.toHaveBeenCalled();
+    expect(editorSpy.revealLineInCenter).toHaveBeenCalledTimes(1);
+    expect(editorSpy.revealLineInCenter).toHaveBeenCalledWith(
+      3,
+      monaco.editor.ScrollType.Smooth
+    );
+    expect(editorSpy.deltaDecorations).toHaveBeenCalledTimes(1);
+
+    // Check the behavior when a new file is focused on.
+    store.overrideSelector(getFocusedSourceFileContent, {
+      loadState: DataLoadState.LOADED,
+      lines: [
+        'model = tf.keras.Sequential',
+        'model.add(tf.keras.layers.Dense(1))',
+      ],
+    });
+    store.overrideSelector(getFocusedSourceLineSpec, {
+      host_name: 'localhost',
+      file_path: '/home/user/model.py',
+      lineno: 1,
+    });
+    store.refreshState();
+    fixture.detectChanges();
+    await fixture.whenStable();
+
+    fileLabelElement = fixture.debugElement.query(By.css('.file-label'));
+    expect(fileLabelElement.nativeElement.innerText).toBe(
+      '/home/user/model.py'
+    );
+    // Switching to a different file relies on setValue() to render the code
+    // of the new file.
+    expect(editorSpy.setValue).toHaveBeenCalledTimes(1);
+    expect(editorSpy.setValue).toHaveBeenCalledWith(
+      'model = tf.keras.Sequential\nmodel.add(tf.keras.layers.Dense(1))'
+    );
+    expect(editorSpy.revealLineInCenter).toHaveBeenCalledTimes(2);
+    expect(editorSpy.revealLineInCenter).toHaveBeenCalledWith(
+      1,
+      monaco.editor.ScrollType.Smooth
+    );
+    expect(editorSpy.deltaDecorations).toHaveBeenCalledTimes(2);
+  });
+
+  it('switching to a different line in the same file', async () => {
+    const fixture = TestBed.createComponent(SourceFilesContainer);
+    store.overrideSelector(getFocusedSourceFileContent, {
+      loadState: DataLoadState.LOADED,
+      lines: ['import tensorflow as tf', '', 'print("hello, world")'],
+    });
+    store.overrideSelector(getFocusedSourceLineSpec, {
+      host_name: 'localhost',
+      file_path: '/home/user/main.py',
+      lineno: 2,
+    });
+    store.refreshState();
+    fixture.detectChanges();
+    await fixture.whenStable();
+
+    // Check the behavior when a new file is focused on.
+    store.overrideSelector(getFocusedSourceLineSpec, {
+      host_name: 'localhost',
+      file_path: '/home/user/main.py',
+      lineno: 1, // Focusing on a different line of the same file.
+    });
+    store.refreshState();
+    fixture.detectChanges();
+    await fixture.whenStable();
```

Comment:
A bit surprised that you need to wait for stability all the time. ~Is this absolutely needed~? Ah, it must be the async/await in the `ngOnChange`.